core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -291,6 +314,12 @@ private[spark] class Executor(
           val reason = cDE.toTaskEndReason
           execBackend.statusUpdate(taskId, TaskState.FAILED, ser.serialize(reason))
 
+        case oom: OutOfMemoryError =>
+          shutdownHookLock.synchronized {
```

Comment:
I don't think we know for *sure* that this thread will always win the race.  If it loses, you dont' want this thread to wait.  I think you can get around this, though, by using [`Lock.tryLock`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/Lock.html#tryLock--) here (and continuing whether or not you get the lock), and using `Lock.lock` from the other thread.